frontend/catalyst/jax_tracer.py

```diff
@@ -1064,58 +1059,69 @@ def trace_quantum_measurements(
-    Let's wait until mid-circuit measurements are better integrated into both PL
-    and Catalyst and discussed more as well."""
-    class_tracers, meas_tracers = split_tracers_and_measurements(flat_results)
-
-    # Can transforms be applied?
-    # Since transforms are a PL feature and PL does not support the same things as
-    # Catalyst, transforms may have invariants that rely on PL invariants.
-    # For example:
-    #   * mid-circuit measurements (for batch-transforms)
-    #   * that the output will be only a sequence of `MeasurementProcess`es.
-    def is_measurement(op):
-        """Only to avoid 100 character per line limit."""
-        return isinstance(op, MeasurementProcess)
+def has_classical_outputs(flat_results):
+    """Checks if the quantum function outputs classical values.
+
+    Returns:
+        bool
+    """
+    for result in flat_results:
+        if not isinstance(result, MeasurementProcess):
+            return True
 
-    is_out_measurements = map(is_measurement, meas_tracers)
-    is_all_out_measurements = all(is_out_measurements) and not class_tracers
-    is_out_measurement_sequence = is_all_out_measurements and isinstance(meas_tracers, Sequence)
-    is_out_single_measurement = is_all_out_measurements and is_measurement(meas_tracers)
+    return False
+
+
+@debug_logger
+def has_midcircuit_measurement(tape):
+    """Check if the tape contains any mid-circuit measurements."""
 
     def is_midcircuit_measurement(op):
         """Only to avoid 100 character per line limit."""
         return isinstance(op, catalyst.api_extensions.MidCircuitMeasure)
 
-    is_valid_output = is_out_measurement_sequence or is_out_single_measurement
-    if not is_valid_output:
-        msg = (
-            "A transformed quantum function must return either a single measurement, "
-            "or a nonempty sequence of measurements."
-        )
-        raise CompileError(msg)
+    return any(map(is_midcircuit_measurement, tape.operations))
+
+
+class TracingMode(Enum):
+    """Enumerate the tracing modes supported by the quantum function tracer:
+
+    DEFAULT - Allows mid-circuit measurements and returns function's results directly.
+              Used when no transform is applied or when transform doesn't modify
+              measurements or produce multiple tapes.
+
+    TRANSFORM - Uses tape measurements instead of original function results.
+                Prohibits mid-circuit measurements with multiple tapes and requires
+                function to return only measurements (no classical results).
+                Used when transform produces multiple tapes or modifies measurements.
+    """
+
+    DEFAULT = 0
+    TRANSFORM = 1
+
+
+@debug_logger
+def have_measurements_changed(original_tape, modified_tape):
+    """Check if the measurement has changed."""
+
+    # Copying tapes preserves object identity in the operation and measurement lists, due to
```

Comment:
Forgot I had this in my other PR:
```suggestion
    if len(original_tape.measurements) != len(modified_tape.measurements):
        return True

    # Copying tapes preserves object identity in the operation and measurement lists, due to
```